File: services/api/internal/model/mfa.go
package model

import (
	"time"

	"github.com/google/uuid"
)

// ---------------------------------------------------------------------------
// Database models
// ---------------------------------------------------------------------------

// TOTPMethod represents a TOTP authenticator app registered by a user.
type TOTPMethod struct {
	ID              uuid.UUID  `json:"-"`
	UserID          uuid.UUID  `json:"-"`
	Name            string     `json:"-"`
	SecretEncrypted []byte     `json:"-"`
	Verified        bool       `json:"-"`
	LastUsedAt      *time.Time `json:"-"`
	CreatedAt       time.Time  `json:"-"`
	UpdatedAt       time.Time  `json:"-"`
}

// WebAuthnCredential represents a WebAuthn security key or passkey.
type WebAuthnCredential struct {
	ID                  uuid.UUID  `json:"-"`
	UserID              uuid.UUID  `json:"-"`
	Name                string     `json:"-"`
	CredentialID        []byte     `json:"-"`
	PublicKey           []byte     `json:"-"`
	AAGUID              []byte     `json:"-"`
	SignCount           int64      `json:"-"`
	Transports          []string   `json:"-"`
	BackupEligible      bool       `json:"-"`
	BackupState         bool       `json:"-"`
	PasswordlessEnabled bool       `json:"-"`
	LastUsedAt          *time.Time `json:"-"`
	CreatedAt           time.Time  `json:"-"`
	UpdatedAt           time.Time  `json:"-"`
}

// RecoveryCode represents a one-time-use recovery code.
type RecoveryCode struct {
	ID        uuid.UUID  `json:"-"`
	UserID    uuid.UUID  `json:"-"`
	CodeHash  string     `json:"-"`
	UsedAt    *time.Time `json:"-"`
	CreatedAt time.Time  `json:"-"`
}

// ServerSetting represents a key-value server setting.
type ServerSetting struct {
	Key       string    `json:"key"`
	Value     string    `json:"value"`
	UpdatedAt time.Time `json:"updated_at"`
}

// ---------------------------------------------------------------------------
// API response types
// ---------------------------------------------------------------------------

// MFAMethodResponse is the JSON representation of an enrolled MFA method.
type MFAMethodResponse struct {
	ID                  uuid.UUID  `json:"id"`
	Type                string     `json:"type"` // "totp" or "webauthn"
	Name                string     `json:"name"`
	Verified            bool       `json:"verified"`
	PasswordlessEnabled bool       `json:"passwordless_enabled,omitempty"`
	LastUsedAt          *time.Time `json:"last_used_at,omitempty"`
	CreatedAt           time.Time  `json:"created_at"`
}

// TOTPSetupResponse is returned when beginning TOTP setup.
type TOTPSetupResponse struct {
	MethodID uuid.UUID `json:"method_id"`
	Secret   string    `json:"secret"`
	URI      string    `json:"uri"`
	Issuer   string    `json:"issuer"`
	Account  string    `json:"account"`
}

// RecoveryCodesResponse is returned when recovery codes are generated.
type RecoveryCodesResponse struct {
	Codes []string `json:"codes"`
}

// MFAChallengeResponse is returned during login when MFA verification is required.
type MFAChallengeResponse struct {
	MFARequired bool     `json:"mfa_required"`
	MFAToken    string   `json:"mfa_token"`
	Methods     []string `json:"methods"` // e.g. ["totp", "webauthn", "recovery"]
}

// ServerSettingsResponse is the JSON representation of server settings.
type ServerSettingsResponse struct {
	MFARequired       bool   `json:"mfa_required"`
	MapboxAccessToken string `json:"mapbox_access_token"`
	GoogleMapsApiKey  string `json:"google_maps_api_key"`
}

// ---------------------------------------------------------------------------
// Request types
// ---------------------------------------------------------------------------

// TOTPSetupRequest is the expected body for beginning TOTP setup.
type TOTPSetupRequest struct {
	Name string `json:"name"`
}

// Validate checks that required fields are present.
func (r *TOTPSetupRequest) Validate() error {
	if r.Name == "" {
		return ErrValidation("name is required")
	}
	return nil
}

// TOTPVerifyRequest is the expected body for verifying a TOTP code during setup.
type TOTPVerifyRequest struct {
	MethodID uuid.UUID `json:"method_id"`
	Code     string    `json:"code"`
}

// Validate checks that required fields are present.
func (r *TOTPVerifyRequest) Validate() error {
	if r.MethodID == uuid.Nil {
		return ErrValidation("method_id is required")
	}
	if r.Code == "" {
		return ErrValidation("code is required")
	}
	if len(r.Code) != 6 {
		return ErrValidation("code must be 6 digits")
	}
	return nil
}

// MFAVerifyTOTPRequest is the expected body for verifying TOTP during login.
type MFAVerifyTOTPRequest struct {
	MFAToken string `json:"mfa_token"`
	Code     string `json:"code"`
}

// Validate checks that required fields are present.
func (r *MFAVerifyTOTPRequest) Validate() error {
	if r.MFAToken == "" {
		return ErrValidation("mfa_token is required")
	}
	if r.Code == "" {
		return ErrValidation("code is required")
	}
	return nil
}

// MFARecoveryRequest is the expected body for using a recovery code during login.
type MFARecoveryRequest struct {
	MFAToken string `json:"mfa_token"`
	Code     string `json:"code"`
}

// Validate checks that required fields are present.
func (r *MFARecoveryRequest) Validate() error {
	if r.MFAToken == "" {
		return ErrValidation("mfa_token is required")
	}
	if r.Code == "" {
		return ErrValidation("code is required")
	}
	return nil
}

// WebAuthnRegisterRequest is the expected body for beginning WebAuthn registration.
type WebAuthnRegisterRequest struct {
	Name string `json:"name"`
}

// Validate checks that required fields are present.
func (r *WebAuthnRegisterRequest) Validate() error {
	if r.Name == "" {
		return ErrValidation("name is required")
	}
	return nil
}

// MFAWebAuthnBeginRequest is the expected body for beginning WebAuthn assertion during login.
type MFAWebAuthnBeginRequest struct {
	MFAToken string `json:"mfa_token"`
}

// Validate checks that required fields are present.
func (r *MFAWebAuthnBeginRequest) Validate() error {
	if r.MFAToken == "" {
		return ErrValidation("mfa_token is required")
	}
	return nil
}

// MFAWebAuthnFinishRequest is the expected body for completing WebAuthn assertion during login.
type MFAWebAuthnFinishRequest struct {
	MFAToken string `json:"mfa_token"`
	// The assertion response is parsed from the raw request body by the handler.
}

// PasskeyToggleRequest is the expected body for toggling passwordless on a credential.
type PasskeyToggleRequest struct {
	PasswordlessEnabled bool `json:"passwordless_enabled"`
}

// UpdateServerSettingsRequest is the expected body for updating server settings.
type UpdateServerSettingsRequest struct {
	MFARequired       *bool   `json:"mfa_required"`
	MapboxAccessToken *string `json:"mapbox_access_token"`
	GoogleMapsApiKey  *string `json:"google_maps_api_key"`
}

// ---------------------------------------------------------------------------
// Redis MFA session data
// ---------------------------------------------------------------------------

// MFASession is stored in Redis during the MFA challenge flow.
type MFASession struct {
	UserID  uuid.UUID `json:"user_id"`
	Methods []string  `json:"methods"`
}

File: clients/ios/SitAware/Models/MFA.swift
import Foundation

// MARK: - MFA Method

/// A single enrolled MFA method (TOTP or WebAuthn).
struct MFAMethod: Codable, Sendable, Identifiable {
    let id: String
    let type: MFAMethodType
    let name: String
    let verified: Bool
    var passwordlessEnabled: Bool?
    var lastUsedAt: String?
    let createdAt: String
}

enum MFAMethodType: String, Codable, Sendable {
    case totp
    case webauthn
}

// MARK: - TOTP Setup

/// Returned when beginning TOTP setup — contains the QR code URI and secret.
struct TOTPSetupResponse: Codable, Sendable {
    let methodId: String
    let secret: String
    let uri: String
    let issuer: String
    let account: String
}

/// Returned when TOTP is verified (first method triggers recovery codes).
struct TOTPVerifyResponse: Codable, Sendable {
    let verified: Bool
    var recoveryCodes: [String]?
}

// MARK: - WebAuthn Registration

/// Returned when WebAuthn registration finishes.
struct WebAuthnRegisterResponse: Codable, Sendable {
    let registered: Bool
    var recoveryCodes: [String]?
}

// MARK: - Recovery Codes

struct RecoveryCodesResponse: Codable, Sendable {
    let codes: [String]
}

// MARK: - Server Settings

struct ServerSettings: Codable, Sendable {
    var mfaRequired: Bool
    var mapboxAccessToken: String
    var googleMapsApiKey: String
}

// MARK: - WebAuthn / Passkey Options (generic JSON containers)

/// Options returned by the server for WebAuthn assertion/registration.
/// We pass these through to ASAuthorization — the exact shape varies, so we
/// decode only the fields we need and keep the rest as raw data.
struct WebAuthnOptions: Codable, Sendable {
    let challenge: String
    var timeout: Int?
    var rpId: String?
    var allowCredentials: [WebAuthnCredentialDescriptor]?
    var userVerification: String?
}

struct WebAuthnCredentialDescriptor: Codable, Sendable {
    let type: String
    let id: String
    var transports: [String]?
}

struct WebAuthnCreationOptions: Codable, Sendable {
    let challenge: String
    let rp: WebAuthnRP
    let user: WebAuthnUser
    var pubKeyCredParams: [WebAuthnPubKeyParam]?
    var timeout: Int?
    var excludeCredentials: [WebAuthnCredentialDescriptor]?
    var authenticatorSelection: WebAuthnAuthenticatorSelection?
    var attestation: String?
}

struct WebAuthnRP: Codable, Sendable {
    let name: String
    let id: String
}

struct WebAuthnUser: Codable, Sendable {
    let name: String
    let displayName: String
    let id: String
}

struct WebAuthnPubKeyParam: Codable, Sendable {
    let type: String
    let alg: Int
}

struct WebAuthnAuthenticatorSelection: Codable, Sendable {
    var authenticatorAttachment: String?
    var residentKey: String?
    var requireResidentKey: Bool?
    var userVerification: String?
}

/// Passkey begin response from the server.
struct PasskeyBeginResponse: Codable, Sendable {
    let options: PasskeyOptionsWrapper
    let sessionId: String
}

/// The server wraps options inside a `publicKey` key.
struct PasskeyOptionsWrapper: Codable, Sendable {
    let publicKey: WebAuthnOptions
}

/// WebAuthn registration begin response.
struct WebAuthnRegisterBeginResponse: Codable, Sendable {
    let publicKey: WebAuthnCreationOptions
}

/// MFA WebAuthn challenge begin response.
struct MFAWebAuthnBeginResponse: Codable, Sendable {
    let options: PasskeyOptionsWrapper
}

// MARK: - WebAuthn Outbound Models (finish requests)

/// Encoder for WebAuthn request bodies.
///
/// The W3C WebAuthn spec requires exact camelCase field names (`clientDataJSON`,
/// `authenticatorData`, `rawId`, etc.), so we cannot use APIClient's default
/// `convertToSnakeCase` encoder. Encode finish request bodies with this encoder
/// and send via `APIClient.postRawJSON`.
enum WebAuthnJSON {
    static let encoder: JSONEncoder = {
        let encoder = JSONEncoder()
        return encoder
    }()
}

/// Assertion response fields (used by passkey login and WebAuthn MFA finish).
struct WebAuthnAssertionResponseData: Encodable, Sendable {
    let authenticatorData: String
    let clientDataJSON: String
    let signature: String
    let userHandle: String?
}

/// Passkey (passwordless) finish request body.
/// Merges `session_id` with the W3C credential assertion fields.
struct PasskeyFinishRequest: Encodable, Sendable {
    let sessionId: String
    let id: String
    let rawId: String
    let type: String
    let response: WebAuthnAssertionResponseData

    enum CodingKeys: String, CodingKey {
        case sessionId = "session_id"
        case id, rawId, type, response
    }
}

/// MFA WebAuthn finish request body.
/// Merges `mfa_token` with the W3C credential assertion fields.
struct MFAWebAuthnFinishRequest: Encodable, Sendable {
    let mfaToken: String
    let id: String
    let rawId: String
    let type: String
    let response: WebAuthnAssertionResponseData

    enum CodingKeys: String, CodingKey {
        case mfaToken = "mfa_token"
        case id, rawId, type, response
    }
}

/// Attestation response fields (used by WebAuthn registration finish).
struct WebAuthnAttestationResponseData: Encodable, Sendable {
    let attestationObject: String
    let clientDataJSON: String
}

/// WebAuthn registration finish request body.
struct WebAuthnRegistrationFinishRequest: Encodable, Sendable {
    let id: String
    let rawId: String
    let type: String
    let response: WebAuthnAttestationResponseData
}

/// Response from passwordless toggle endpoint.
struct PasswordlessToggleResponse: Decodable, Sendable {
    let passwordlessEnabled: Bool
}
